### PR TITLE
Split deserialize from runPythonUdf and remove generatePythonUDFResult

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34497 Don't run user function until all UserRRefs in the args are confirmed
* #34809 Fix Error Hanlding for Local Send
* **#34496 Split deserialize from runPythonUdf and remove generatePythonUDFResult**

Differential Revision: [D20347469](https://our.internmc.facebook.com/intern/diff/D20347469)

### CODE REVIEW

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 1]
```diff
@@ -70,11 +70,17 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
     }
     case MessageType::PYTHON_CALL: {
       auto& pyCall = static_cast<PythonCall&>(rpc);
-      auto serializedPyObj =
-          PythonRpcHandler::getInstance().generatePythonUDFResult(
-              pyCall.serializedPyObj());
+      auto& pythonRpcHandler = PythonRpcHandler::getInstance();
+      std::shared_ptr<SerializedPyObj> serializedPyObj = nullptr;
+      {
+        pybind11::gil_scoped_acquire ag;
```
Comment (zhaojuanmao) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: do we need GIL here?  serialize and deserialize already hold a GIL inside

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 2]
```diff
@@ -140,9 +146,15 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
...
-      IValue py_ivalue = jit::toIValue(
-          PythonRpcHandler::getInstance().runPythonUDF(prc.serializedPyObj()),
-          PyObjectType::get());
+      auto& pythonRpcHandler = PythonRpcHandler::getInstance();
+      IValue py_ivalue;
+      {
+        pybind11::gil_scoped_acquire ag;
+        auto pythonUdf = pythonRpcHandler.deserialize(prc.serializedPyObj());
+        py_ivalue = jit::toIValue(
+            PythonRpcHandler::getInstance().runPythonUdf(std::move(pythonUdf)),
```
Comment (zhaojuanmao) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: same as above, does it need GIL here?
pythonRpcHandler.runPythonUdf()

Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: When I drop the GIL, I hit errors/hangs in tests. Looks like even std::move an `py::object` requires GIL?

**Review by pritamdamania87 — COMMENTED** [review]
Looks good, although CI says there was a failure introduced in this PR.

**Review by rohan-varma — APPROVED** [review]
Stamp, although please make sure tests pass

**Pull request closed without merging**